content/chapters/io/lecture/slides/IPC.md

```diff
@@ -0,0 +1,189 @@
+### Interprocess Communication
+
+---
+
+### Interprocess Communication
+
+- A communication channel involves a process and a data endpoint
+- We can use persistent data and I/O devices as endpoints using the **file interface**
+- Another possible endpoint is a **process**
+  - Law of the hammer
+  - Socket interface
+
+----
+
+### IPC - File Interface
+
+![IPC through File Interface](./media/IPC-file-interface.svg)
+
+- Writing to disk and reading from disk is tedious
+
+----
+
+### IPC - Pipe
+
+![IPC through pipe](./media/IPC-pipe.svg)
+
+- Same idea as above
+
+----
+
+- **Proc1** writes at one end and **Proc2** read from the other end
+  - Why is it better?
+  - The **pipe** is stored in Kernel Space, we do not work directly with the disk
+- Only works for related processes ([fork](https://man7.org/linux/man-pages/man2/fork.2.html))
+
+----
+
+### `pipe()` demo
+
+- `demo/IPC/pipe.c`
+
+----
+
+### IPC - Named Pipe (FIFO)
+
+- Bypass unnamed pipes limitations:
+  - Can be used by more than 2 processes
+  - Can be used by unrelated processes
+- Stored on disk as a **special file**
+
+```console
+student@OS:~$ ls -l my_fifo
+prw-r--r-- 1 student student 0 nov 22 18:38 my_fifo
+```
+
+----
+
+### `mkfifo()` demo
+
+- `demo/IPC/fifo.c`
+
+---
+
+### IPC - Socket Interface
+
+![IPC through Socket](./media/IPC-socket.svg)
+
+----
+
+- Designed for interprocess communication
+- Communication between processes on the **same host** or on **different hosts**
+- Supports multiple transport protocols
+  - E.g.: `TCP`, `UDP`
+
+----
+
+### Networking
+
+- Each host is identified by an [**IP address**](https://ocw.cs.pub.ro/courses/uso/laboratoare/laborator-05)
+- For a host to use multiple applications that require network connection we use **ports**
+  - A port is an integet between **0** and **65536**
+- An application is identified by an IP address and a port
+  - E.g.: `217.182.27.243:25565`
+
+----
+
+### Client and Server
+
+- A process that use the socket interface is either a client or a server
+  - The server waits for messages/connections on IP and port
+  - The client attempts to connect to a server using IP and port
+
+----
+
+### TCP and UDP
```

Comment:
I would use the terms `Stream` and `Datagram`, and map them to `SOCK_STREAM` and `SOCK_DGRAM`.

Reply:
Agreed.